### PR TITLE
libs/libxx: uClibc depends on CONFIG_ALLOW_GPL_COMPONENTS=y

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary
libs/libxx: uClibc depends on CONFIG_ALLOW_GPL_COMPONENTS=y
uClibc is LGPL so it should depends on CONFIG_ALLOW_GPL_COMPONENTS=y

https://www.uclibc.org/FAQ.html#licensing

## Impact
LGPL code must be protected by CONFIG_ALLOW_GPL_COMPONENTS option.

## Testing
none

### CODE REVIEW

**Comment by acassis** [comment]
@raiden00pl GPL and LGPL are different licenses, should we create a CONFIG_ALLOW_LGPL_COMPONENTS for this case?

File: libs/libxx/Kconfig [hunk 1]
```diff
@@ -53,6 +53,7 @@ config LIBCXX
 config UCLIBCXX
 	bool "uClibc++ Standard C++ Library"
 	select HAVE_CXXINITIALIZE
+	depends on ALLOW_GPL_COMPONENTS
```
Comment (xiaoxiang781216) [inline libs/libxx/Kconfig:56]: should be change to LGPL

**Comment by raiden00pl** [comment]
ALLOW_GPL_COMPONENTS is for both: GPL and LGPL:

```
config ALLOW_GPL_COMPONENTS
	bool "Use components that have GPL/LGPL licenses"
	default n
```

from ASF perspective both are in the same category: https://www.apache.org/legal/resolved.html#category-x
We don't use uClibc as linked library but compile its code directly and there are some controversy about object-oriented programming and LGPL-licensed code (which was solved in LGPL v3, but this code is LGPL v2.1), so I don't know if LGPL has any meaning here at all.

I think it's safer to keep copyleft licenses as one option, but maybe someone who knows something more about licenses has a different opinion.

**Comment by acassis** [comment]
> ALLOW_GPL_COMPONENTS is for both: GPL and LGPL:
> 
> ```
> config ALLOW_GPL_COMPONENTS
> 	bool "Use components that have GPL/LGPL licenses"
> 	default n
> ```
> 
> from ASF perspective both are in the same category: https://www.apache.org/legal/resolved.html#category-x We don't use uClibc as linked library but compile its code directly and there are some controversy about object-oriented programming and LGPL-licensed code (which was solved in LGPL v3, but this code is LGPL v2.1), so I don't know if LGPL has any meaning here at all.
> 
> I think it's safer to keep copyleft licenses as one option, but maybe someone who knows something more about licenses has a different opinion.

I'm not license expert either, but...

In theory LGPL you don't need to release the source code, even with statically linked (but you need to release the file objects to be linked with new library version). But again this is a grey area, so maybe it is better keep GPL/LGPL as a GPL, this way the users will be aware of the issue.

Maybe we should consider printing a message before the nsh> show up to let users know that the code is linked against LGPL and the source code need to be released case the product is used for commercial application

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by raiden00pl** [comment]
@acassis keeping it as a GPL dependency is a defensive move to protect the user. If the code is LGPL then we leave it up to the user to determine what the resulting license for their project will be. In the case of copyleft licenses, it's better to assume the worst case scenario.

**Comment by jerpelea** [comment]
@raiden00pl 
I think that the GPL components should not be placed in NuttX git, they should be moved to the apps git 
GPL components should never be enabled by default because there is a high risk that they are overlooked.

**Comment by raiden00pl** [comment]
there are 3 types of GPL configs in NuttX now:

1. socketCAN related apps - but these components are dual licence (GPL-2.0-only OR BSD-3-Clause) and I don't understand why they depend on GPL at all. We can use BSD-3 for them and then we don't have dependency on GPL

2. uClibc which is LGPL and there's no doubt about it. We can remove these configurations, then we lose coverage in CI, or warn the user in some other visible way (maybe big warning during compilation?)

And another thing is `drivers/video/ov2640.c` file, which I have no idea how to hadle.

**Comment by acassis** [comment]
> @jerpelea there are 3 types of GPL configs in NuttX now:
> 
>     1. socketCAN related apps - but these components are dual licence (GPL-2.0-only OR BSD-3-Clause) and I don't understand why they depend on GPL at all. We can use BSD-3 for them and then we don't have dependency on GPL
> 
>     2. uClibc which is LGPL and there's no doubt about it. We can remove these configurations, then we lose coverage in CI, or warn the user in some other visible way (maybe big warning during compilation?)
> 
> 
> And another thing is `drivers/video/ov2640.c` file, which I have no idea how to hadle.

1. Agreed, should be moved to BSD COMPONENT
2. An alternative is using a bit flag, everything some GPL/LGPL code is compiled/linked it is enabled and a message could be printed just before the NSH start. A warning during the compilation could be easy missed
3. I think ov2640 is not GPL code at all, as I remember @patacongo just looked the initialization table that was from some GPL driver, but AFAIK Omnivision supply it as sample code as well

**Comment by xiaoxiang781216** [comment]
CONFIG_xxx can be got from:

1. elf debug info or
2. https://github.com/apache/nuttx-apps/tree/master/system/nxdiag
so, don't need invent another method to get the same information again.

**Comment by raiden00pl** [comment]
> An alternative is using a bit flag, everything some GPL/LGPL code is compiled/linked it is enabled and a message could be printed just before the NSH start. A warning during the compilation could be easy missed

The problem with this approach is that nuttx-apps and NSH is optional, not all users use them. Information like this should be displayed to every user, so it should be on the kernel side

**Pull request merged**